src/Authentication/Authenticators/Session.php

```diff
@@ -88,9 +93,59 @@ public function attempt(array $credentials): Result
 
         $this->recordLoginAttempt($credentials, true, $ipAddress, $userAgent, $user->getAuthId());
 
+        // If an action has been defined for login, start it up.
+        $actionClass = setting('Auth.actions')['login'] ?? null;
+        if (! empty($actionClass)) {
+            session()->set('auth_action', $actionClass);
+        } else {
+            // a successful login
+            Events::trigger('login', $user);
+        }
+
         return $result;
     }
 
+    /**
+     * Check token in Action
+     *
+     * @param string $type  Action type. 'email_2fa' or 'email_activate'
+     * @param string $token Token to check
+     */
+    public function checkAction(string $type, string $token): bool
+    {
+        $user = $this->loggedIn()
+            ? $this->getUser()
+            : null;
+
+        $identity = $user->getIdentity($type);
+
+        if (empty($token) || $token !== $identity->secret) {
+            return false;
+        }
+
+        /** @var UserIdentityModel $identityModel */
+        $identityModel = model(UserIdentityModel::class);
+
+        // On success - remove the identity and clean up session
+        $identityModel->deleteIdentitiesByType($user->getAuthId(), $type);
+
+        // Clean up our session
+        session()->remove('auth_action');
+
+        // a successful login
+        Events::trigger('login', $user);
```

Comment:
Would this always be a login event?

Reply:
Looks like there's no event fired after EmailActivation.

Reply:
The method is called in the action `Email2FA` or `EmailActivator`.
The `login` event is fired after token checking.
The event is fired after EmailActivation.

Reply:
The `login` should only be fired when a user manually logged in. Just checking the token (like a remember-me check) should not fire it off. 

Also, if the user has activated their email, the login action will fire the event for them. No need to do it here. If we still need it fired for the `Email2FA` action I'd move the event out to that class.

Reply:
In this case, the login event is fired after 2FA token checking.

>  If we still need it fired for the Email2FA action I'd move the event out to that class.

This seems to be a major point of disagreement.

I think the Action classes (`Email2FA`) are parts of controllers, and controllers are not a part of Auth library.
But Authenticators are a part of Auth library.
Auth library manages authentications and should fire login events.

So I removed `Events::trigger()` in controllers as much as possible.

Reply:
That's fair, but login doesn't seem appropriate after EmailActivation. And that method covers all actions, doesn't it? So it would fire login after 2FA, but also after registration, which is not desirable. Especially since the `login` event is now triggered in the `login()` method itself. So this seems unnecessary, is all I'm saying. Am I understanding the flow correctly?

Reply:
> And that method covers all actions, doesn't it?

Yes.

> So it would fire login after 2FA, but also after registration, which is not desirable.

Yes. I think to fire login event after registration is okay. 

> Same thing applies to the email activation process.  Users are automatically logged in to the application after verifying their email address. The didLogin event should also be fired. https://github.com/codeigniter4/shield/pull/99

But if it should not, I will skip the `Events::trigger()` when 'email_activate'.

> Especially since the login event is now triggered in the login() method itself.

I removed the trigger in `login()` method in this PR. 
But the issue is what `login()` method does, what `login` is.

If `login` means authentication is complete, i.e. the user has no further Auth steps to take (see https://github.com/codeigniter4/shield/issues/128#issuecomment-1126691010),
and the `login()` method makes a user state so, the `login` event should be fired in it.

Reply:
> I removed the trigger in login() method in this PR.

You're right. I missed that. 

I think I'm convinced enough to allow that to stay. And I agree with a pending state until complete authorization.